### PR TITLE
ARROW-12592: [C++][Gandiva] Add init_cap function

### PR DESCRIPTION
**Pull request by anthonylouisbsb**
The function capitalizes the first letter in each word of a given text.

### CODE REVIEW

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 1]
```diff
@@ -386,6 +389,142 @@ GDV_FN_CAST_VARCHAR_REAL(float64, DoubleType)
...
+
+    bool is_char_space = gdv_fn_is_codepoint_for_space(char_codepoint);
+
+    int32_t formatted_codepoint;
+    if (last_char_was_space && !is_char_space) {
+      // Convert the encoded codepoint to its uppercase codepoint
+      formatted_codepoint = utf8proc_toupper(char_codepoint);
+    } else {
+      // Leave
+      formatted_codepoint = static_cast<int32_t>(char_codepoint);
```
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs.cc:509]: why extra cast here

**Review by projjal — COMMENTED** [review]
Can you include this patch into the other upper/lower PR, since a lot of the changes are same

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 1 above)
Comment (anthonylouisbsb) [reply cpp/src/gandiva/gdv_function_stubs.cc:509]: cast removed, I will put the PR changes in the upper/lower PR

**Comment by anthonylouisbsb** [comment]
The PR will be closed and all the changes will be added in the https://github.com/apache/arrow/pull/10040 PR

**Pull request closed without merging**